### PR TITLE
*: reduce mem allocs caused by GetTextToLog

### PR DESCRIPTION
**Pull request by zyguan**
### What problem does this PR solve?

Issue Number: ref #54047

Problem Summary: Many memory allocations caused by `GetTextToLog` are unnecessary. Typically, we only need the sql text when: 1. record stmt summary for commit (which needs sql text of prev stmt) 2. new stmt record (when the LRU cache miss).

### What changed and how does it work?

Evaluate sql text lazily. Refer to the following benchmark results of sysbench oltp point select.

https://clinic.pingcap.com.cn/portal/#/orgs/541/clusters/7381654358775158872?from=1718790105&to=1718790705

  | avg lat | p95 lat | throughput | througput diff
-- | -- | -- | -- | --
baseline | 1.52 | 3.02 | 65740.95 | -
this PR | 1.49 | 2.91 | 67045.73 | 1.98%

baseline
image

this PR
image

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/sessionctx/variable/session.go [hunk 1]
```diff
@@ -1926,6 +1927,48 @@ func (p *PlanCacheParamList) AllParamValues() []types.Datum {
...
+type LazyStmtText struct {
+	text   *string
+	SQL    string
+	Redact string
+	Params PlanCacheParamList
+	Format func(string) string
+}
+
+// SetText sets the text directly.
+func (s *LazyStmtText) SetText(text string) {
```
Comment (cfzjywxk) [inline pkg/sessionctx/variable/session.go:1940]: How about 
```
func (s *LazyStmtText) SetText(text *string) {
   s.text = text
}
```

File: pkg/executor/adapter.go [hunk 2]
```diff
@@ -2023,6 +2025,43 @@ func (a *ExecStmt) GetTextToLog(keepHint bool) string {
 	return sql
 }
 
+// getLazyText is equivalent to `a.GetTextToLog(false)`. Note that the s.Params is a shallow copy of
+// `sessVars.PlanCacheParams`, so you can only use the lazy text within the current stmt context.
+func (a *ExecStmt) getLazyStmtText() (s variable.LazyStmtText) {
+	sessVars := a.Ctx.GetSessionVars()
+	rmode := sessVars.EnableRedactLog
+	if rmode == errors.RedactLogEnable {
+		sql, _ := sessVars.StmtCtx.SQLDigest()
```
Comment (cfzjywxk) [inline pkg/executor/adapter.go:2034]: Does L2034 and L2037 introduce memory allocation and copy, and is there any optimization space for these two functions called?

File: pkg/sessionctx/variable/session.go (hunk 1 above)
Comment (zyguan) [reply pkg/sessionctx/variable/session.go:1940]: Not much difference I think.

**Review by cfzjywxk — APPROVED** [review]
LGTM

**Comment by cfzjywxk** [comment]
@easonn7
PTAL, the session var file is modified but there is no configuration related changes.

**Review by crazycs520 — APPROVED** [review]

**Comment by cfzjywxk** [comment]
@easonn7
PTAL, the variable related file is modifed but there is no changes for system variables.

**Pull request merged**